core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/UpdatedColumnDecoder.scala

```diff
@@ -139,56 +129,36 @@ abstract class UpdatedColumnDecoderBase(decoder: ColumnDecoder, field: StructFie
         movedIndex = 1
       }
     }
-    // last delta in hierarchy
-    if (delta3Position < next) {
-      // skip on equality (result should be returned by one of the previous calls)
-      if (delta3Position <= ordinal) {
-        skipUpdatedPosition(delta3)
-        delta3Position = delta3.moveToNextPosition()
-        if (delta3Position < next) {
-          next = delta3Position
-          movedIndex = 2
-        }
-      } else {
-        next = delta3Position
-        movedIndex = 2
-      }
-    }
     movedIndex match {
       case 0 =>
         delta1Position = delta1.moveToNextPosition()
         nextDeltaBuffer = delta1
       case 1 =>
         delta2Position = delta2.moveToNextPosition()
         nextDeltaBuffer = delta2
-      case 2 =>
-        delta3Position = delta3.moveToNextPosition()
-        nextDeltaBuffer = delta3
       case _ =>
     }
     next
   }
 
   private def skipUntil(ordinal: Int): Boolean = {
     var nextUpdated = nextUpdatedPosition
-    // check if ordinal has moved ahead of updated cursor
-    if (nextUpdated < ordinal) {
-      do {
-        // skip the position in current delta
-        skipUpdatedPosition(nextDeltaBuffer)
-        // update the cursor and keep on till ordinal is not reached
-        nextUpdated = moveToNextUpdatedPosition(nextUpdated)
-      } while (nextUpdated < ordinal)
-      nextUpdatedPosition = nextUpdated
-      if (nextUpdated > ordinal) return true
-    }
+    do {
+      // skip the position in current delta
+      skipUpdatedPosition(nextDeltaBuffer)
+      // update the cursor and keep on till ordinal is not reached
+      nextUpdated = moveToNextUpdatedPosition(nextUpdated)
+    } while (nextUpdated < ordinal)
+    nextUpdatedPosition = nextUpdated
+    if (nextUpdated > ordinal) return true
     currentDeltaBuffer = nextDeltaBuffer
```

Comment:
Do currentDeltaBuffer and nextUpdatedPosition need to be updated if (nextUpdatedPosition == ordinal)?

Reply:
This will change and nextDeltaBuffer is being removed. There will only be currentDeltaBuffer which will be updated when ordinal moves past nextUpdatedPosition so only currentDeltaBuffer will be required.